actors/builtin/market/market_test.go

```diff
@@ -1 +1,252 @@
+
+import (
+	"context"
+	"testing"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+	"github.com/filecoin-project/specs-actors/actors/builtin/market"
+	"github.com/filecoin-project/specs-actors/actors/builtin/miner"
+	"github.com/filecoin-project/specs-actors/actors/runtime/exitcode"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"github.com/filecoin-project/specs-actors/support/mock"
+	tutil "github.com/filecoin-project/specs-actors/support/testing"
+)
+
+func TestActorConstruction(t *testing.T) {
+	actor := market.Actor{}
+	receiver := tutil.NewIDAddr(t, 100)
+	builder := mock.NewBuilder(context.Background(), receiver).
+		WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+	t.Run("simple construction", func(t *testing.T) {
+		rt := builder.Build(t)
+
+		rt.ExpectValidateCallerAddr(builtin.SystemActorAddr)
+
+		params := adt.EmptyValue{}
+		ret := rt.Call(actor.Constructor, &params).(*adt.EmptyValue)
+
+		assert.Equal(t, adt.EmptyValue{}, *ret)
+		rt.Verify()
+	})
+}
+
+func TestAddBalance(t *testing.T) {
+	marketActor := tutil.NewIDAddr(t, 100)
+	provider := tutil.NewIDAddr(t, 101)
+	worker := tutil.NewIDAddr(t, 102)
+	client := tutil.NewIDAddr(t, 103)
+
+	var st market.State
+
+	setup := func() (*mock.Runtime, *marketActorTestHarness) {
+		builder := mock.NewBuilder(context.Background(), marketActor).
+			WithBalance(abi.NewTokenAmount(100), abi.NewTokenAmount(0)).
+			WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+		rt := builder.Build(t)
+
+		actor := marketActorTestHarness{t: t}
+		actor.constructAndVerify(rt)
+
+		return rt, &actor
+	}
+
+	t.Run("adds to provider escrow funds", func(t *testing.T) {
+		rt, actor := setup()
+
+		testCases := []struct {
+			delta int64
+			total int64
+		}{
+			{10, 10},
+			{20, 30},
+			{40, 70},
+		}
+
+		for _, tc := range testCases {
+			rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+			rt.SetReceived(abi.NewTokenAmount(tc.delta))
+			actor.expectProviderControlAddress(rt, provider, worker)
+
+			rt.Call(actor.AddBalance, &provider)
+
+			rt.Verify()
+
+			rt.GetState(&st)
+			assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, provider))
+		}
+	})
+
+	t.Run("adds to non-provider escrow funds", func(t *testing.T) {
+		rt, actor := setup()
+
+		testCases := []struct {
+			delta int64
+			total int64
+		}{
+			{10, 10},
+			{20, 30},
+			{40, 70},
+		}
+
+		for _, tc := range testCases {
+			rt.SetCaller(client, builtin.AccountActorCodeID)
+			rt.SetReceived(abi.NewTokenAmount(tc.delta))
+			rt.ExpectValidateCallerType(builtin.CallerTypesSignable...)
+
+			rt.Call(actor.AddBalance, &client)
+
+			rt.Verify()
+
+			rt.GetState(&st)
+			assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, client))
+		}
+	})
```

Comment:
Please add a case for the somewhat confusing possibility of a miner worker account functioning as the client of a deal (i.e. the worker account address as the parameter), ensuring the balance is credited to the worker account, not the miner account.